### DISCUSSION TITLE
When specifying dedicated gpu, Failed to locate an output device

### DISCUSSION THREAD

**Discussion by skylestia** [LizardByte/community#260] [Sunshine] [answered]
Hi! I've been trying to figure out how to tell Sunshine to use my dedicated GPU for a while but just haven't been able to get it to work. I have an MSI GL65 Leopard gaming laptop that I'm trying to use as a host. It has an Intel i7-10750H and an NVIDIA RTX 2070. It has both an integrated GPU and a dedicated GPU.

I'm on the nightly release on Sunshine Version 0.20.0.8ff2022aaf7c4d19b57f505df9b7d7e1b6060365 and I also experience this issue on the stable release

Running dxgi-info.exe returns:

```
>dxgi-info.exe
====== ADAPTER =====
Device Name      : NVIDIA GeForce RTX 2070
Device Vendor ID : 0x000010DE
Device Device ID : 0x00001F14
Device Video Mem : 8019 MiB
Device Sys Mem   : 0 MiB
Share Sys Mem    : 8102 MiB

    ====== OUTPUT ======
    Output Name       : \\.\DISPLAY1
    AttachedToDesktop : yes
    Resolution        : 1920x1080

====== ADAPTER =====
Device Name      : Intel(R) UHD Graphics
Device Vendor ID : 0x00008086
Device Device ID : 0x00009BC4
Device Video Mem : 128 MiB
Device Sys Mem   : 0 MiB
Share Sys Mem    : 8102 MiB

    ====== OUTPUT ======
====== ADAPTER =====
Device Name      : Microsoft Basic Render Driver
Device Vendor ID : 0x00001414
Device Device ID : 0x0000008C
Device Video Mem : 0 MiB
Device Sys Mem   : 0 MiB
Share Sys Mem    : 8102 MiB

    ====== OUTPUT ======
```

In Sunshine's settings I set the Adapter Name to `NVIDIA GeForce RTX 2070` and tried leaving Output Name blank as well as setting it to `\\.\DISPLAY1`. If I specify NVIDIA GeForce RTX 2070, opening Sunshine returns `Error: Failed to locate an output device` for every single encoder and Moonlight fails to connect. Letting Sunshine choose a GPU automatically lets me connect, but it always uses the integrated GPU. Specifying the Output Name or not has no effect.

Based on various threads and discussions I found on Github and Reddit while Googling this issue, I have also tried these with no effect:

- set the NVIDIA dedicated GPU as the preferred graphics processor and set it as the dedicated PhysX processor in the NVIDIA control panel. 
- told Windows to use high performance NVIDIA GPU for sunshine
- tried disabling the integrated GPU

Thanks!

**Comment by ReenigneArcher** [comment, marked as answer]
On Windows the display you want to stream must be connected to the selected GPU.

**Reply by skylestia** [reply]
Isn't it connected to the GPU? How can I tell if it's not? If I understand how to read the output of dxgi-info.exe it looks to me like it is attached to the GPU, but I'm not positive.

Also, do you know what I could do to change it so that the display **_is_** using the GPU instead of the CPU? I went as far as to disable the integrated GPU in device manager and Sunshine would still not use my dedicated GPU. So idk what to do or search for now.

**Reply by ReenigneArcher** [reply]
I could be wrong, but I think most laptops connect the display to the integrated GPU, not the dedicated GPU.

You can't really change it or do anything about it.

**Reply by skylestia** [reply]
Oh! that sucks lol Thanks anyway

I'm not really having any issues with Sunshine using the integrated GPU in terms of performance or anything, I just wanted to use HEVC which is not working on the iGPU for me. But I thought it must be that my iGPU just doesn't support it where my dedicated GPU does cuz I know that OBS for example can use HEVC if I stream online. idk whether OBS is using the integrated GPU or not though. I also don't know how different streaming online is to streaming in-home lol

But it's not that important really, since Sunshine performs so well on my iGPU anyway. It's way, way better for me than Steam Link or any other solution I've ever tried!

Thanks again! 👋

**Comment by n0routine** [comment]
The information here settled a lot of the weird stuff I experienced figuring this out, thanks.

> I could be wrong, but I think most laptops connect the display to the integrated GPU, not the dedicated GPU.
> You can't really change it or do anything about it.

Some addendum to this that I feel relevant:

While the internal display is often connected to the IGPU, I wondered if the HDMI video-out would be connected to the DGPU, since on PCs that's where you find HDMI ports. Unfortunately, it doesn't seem to be the case on my gaming laptop. According to the output from `dxgi-info.exe`. whatever I set my external display mode to (mirror, extend, secondary only), it was being rendered by the IGPU (I checked using a $2 HDMI dummy plug for convenience, but I see no reason the result would be different with an actual monitor). Apparently the HDMI port is connected directly to the IGPU, with the DGPU taking a render-only role for applications to use it and forward the pixels to the IGPU. 

Now, finding reliable information about any non-obvious windows behaviour is a headache, thanks in no small part to microsoft's high-quality answers on its own support forums where the enlightening solution to every single problem ever is either to reinstall drivers, reinstall windows, or a link to some 3rd party site helpfully teaching you how to plug in an HDMI cable.

But I believe things might be different across different laptops, where the HDMI port *might* work as a solution to this problem. Some laptops might have a setting to change this in their UEFI. Certain laptops these days (ASUS I believe?) actually come with a electronic muxer that can control which GPU runs the display device at a hardware level.

**Comment by n0routine** [comment]
> idk whether OBS is using the integrated GPU or not though.

According to their docs, OBS can use both GPUs. But the condition remains that the capturing must happen on the same GPU where the rendering is happening. So apparently desktop-level capturing is independent from application-level capturing. That is, you can use your DGPU to capture a game that is running on your DGPU (and similarly the IGPU for IGPU applications), but if you want to capture your entire (fully stitched and rendered) desktop, you have to use the GPU that is rendering your desktop, which is probably your IGPU.

This all check out, considering I've been using the desktop mode in moonlight to test everything. And I do really need the desktop mode, as I can use it as a general purpose remote desktop.

But this would also imply that the behaviour should be different if you launch a DGPU game from sunshine for streaming, as it will only need to stream the DGPU surface? I do wonder...

**Comment by skylestia** [comment]
> While the internal display is often connected to the IGPU, I wondered if the HDMI video-out would be connected to the DGPU, since on PCs that's where you find HDMI ports. Unfortunately, it doesn't seem to be the case on my gaming laptop. According to the output from `dxgi-info.exe`. whatever I set my external display mode to (mirror, extend, secondary only), it was being rendered by the IGPU (I checked using a $2 HDMI dummy plug for convenience, but I see no reason the result would be different with an actual monitor). Apparently the HDMI port is connected directly to the IGPU, with the DGPU taking a render-only role for applications to use it and forward the pixels to the IGPU.

Thanks for the info! In some of my searches I saw that many newer gaming laptops come equipped with a piece of hardware called a MUX switch. If I understand correctly, a MUX switch is a physical switch somewhere in the computer that controls which GPU gets used. It looks like I bought my laptop *just a couple months* too early to have been in the time range where it started becoming more common for gaming laptops to have it. My laptop is supposedly supposed to be very extensible and moddable which is one of the reasons I bought it, so I've been thinking about looking into whether it's possible to change the way the GPUs are setup in mine, but it might be too much effort for me.

Personally, my use case for Sunshine is a system where I'm treating my laptop more or less like a console for various other devices in my home: I keep it tucked away with the screen brightness all the way down and the keyboard backlight off on a shelf near my living room TV and use a game library manager called Playnite in fullscreen to act as the primary interface for the "console." It's loaded up with a bunch of old emulated games and newer games from Steam and GOG and stuff, I even configured some TV streaming apps in it! So I'm trying to keep the actual physical form factor down so it can fit neatly on its shelf so an external monitor wouldn't really be feasible for my purposes anyway.

> But this would also imply that the behaviour should be different if you launch a DGPU game from sunshine for streaming, as it will only need to stream the DGPU surface? I do wonder...

That is very interesting, I may test it later. Unfortunately I'd like to keep running Sunshine capturing my desktop rather than games so that I can keep treating Playnite as the primary interface for launching games (it's much prettier (with a custom skin) than Moonlight's launcher interface xD) but I may consider it worth it to reconsider the setup if Sunshine can use the dedicated GPU if it's capturing a game running on it. Or maybe I can just have Moonlight launch Playnite and have Windows run Playnite on the dGPU. Although I'm not sure if Sunshine would capture any games that Playnite launches in that case... I'll have to test and see.
